protocol-designer/src/step-forms/utils/index.ts

```diff
@@ -248,3 +296,84 @@ export function getHydratedForm(
   // @ts-expect-error(sa, 2021-6-14):type this properly in #3161
   return hydratedForm
 }
+
+export function getUnoccupiedSlotForMoveableTrash(
+  file: PDProtocolFile,
+  hasWasteChuteCommands: boolean,
+  stagingAreaSlotNames: AddressableAreaName[]
+): string {
+  const wasteChuteSlot = hasWasteChuteCommands
+    ? [WASTE_CHUTE_CUTOUT as string]
+    : []
+  const stagingAreaCutoutIds = stagingAreaSlotNames.map(slotName =>
+    getCutoutIdByAddressableArea(
+      slotName,
+      'stagingAreaRightSlot',
+      FLEX_ROBOT_TYPE
+    )
+  )
+  const allLoadLabwareSlotNames = Object.values(file.commands)
+    .filter(
+      (command): command is LoadLabwareCreateCommand =>
+        command.commandType === 'loadLabware'
+    )
+    .reduce((acc: string[], command) => {
+      const location = command.params.location
+      if (
+        location !== 'offDeck' &&
+        location !== null &&
+        'slotName' in location
+      ) {
+        return [...acc, location.slotName]
+      }
+      return acc
+    }, [])
+
+  const allLoadModuleSlotNames = Object.values(file.commands)
+    .filter(
+      (command): command is LoadModuleCreateCommand =>
+        command.commandType === 'loadModule'
+    )
+    .flatMap(command => {
+      //  special-casing Thermocycler
+      if (command.params.model === THERMOCYCLER_MODULE_V2) {
+        return ['A1', command.params.location.slotName]
+      } else {
+        return command.params.location.slotName
+      }
+    })
+
+  const allMoveLabwareLocations = Object.values(file.commands)
+    .filter(
+      (command): command is MoveLabwareCreateCommand =>
+        command.commandType === 'moveLabware'
+    )
+    .reduce((acc: string[], command) => {
+      const newLocation = command.params.newLocation
+      if (
+        newLocation !== 'offDeck' &&
+        newLocation !== null &&
+        'slotName' in newLocation
+      ) {
+        return [...acc, newLocation.slotName]
+      }
+      return acc
+    }, [])
+
+  const unoccupiedSlot = MOVABLE_TRASH_CUTOUTS.find(
+    cutout =>
+      !allLoadLabwareSlotNames.includes(cutout.slot) &&
+      !allLoadModuleSlotNames.includes(cutout.slot) &&
+      !allMoveLabwareLocations.includes(cutout.slot) &&
+      !wasteChuteSlot.includes(cutout.value) &&
+      !stagingAreaCutoutIds.includes(cutout.value as CutoutId)
+  )
+  if (unoccupiedSlot == null) {
+    console.error(
+      'Expected to find an unoccupied slot for auto-generating a trash bin but could not'
+    )
+    return ''
+  }
+
+  return unoccupiedSlot?.slot
```

Comment:
this line is after null-check so `?` is not required.

```suggestion
  return unoccupiedSlot.slot
```